pyphare/pyphare_tests/test_pharesee/test_geometry.py

```diff
@@ -0,0 +1,528 @@
+     - refinement_boxes
+     """
+    if "simulation" in kwargs:
+        for k in kwargs:
+            if k != "simulation":
+                print("warning: 'simulation' given, {} discarded".format(k))
+
+        sim = kwargs["simulation"]
+        nbr_cells = sim.cells[0]
+        origin = sim.origin[0]
+        interp_order = sim.interp_order
+        domain_size = sim.simulation_domain()[0]
+        cell_width = sim.dl[0]
+        refinement_ratio = 2
+        refinement_boxes = sim.refinement_boxes
+
+    else:
+        nbr_cells = kwargs["nbr_cells"]
+        origin = kwargs["origin"]
+        interp_order = kwargs["interp_order"]
+        domain_size = kwargs["domain_size"]
+        cell_width = kwargs["cell_width"]
+        refinement_ratio = kwargs.get("refinement_ratio", 1)
+        refinement_boxes = kwargs.get("refinement_boxes", {})
+
+    domain_box = boxm.Box(0, nbr_cells - 1)
+    domain_layout = GridLayout(domain_box, origin, cell_width, interp_order)
+
+    coarse_particles = Particles(box=domain_box)
+
+    # copy domain particles and put them in ghost cells
+    particle_ghost_nbr = domain_layout.particleGhostNbr(interp_order)
+    box_extend = particle_ghost_nbr - 1
+
+    upper_slct_box = Box(domain_box.upper - box_extend, domain_box.upper)
+    lower_slct_box = Box(domain_box.lower, domain_box.lower + box_extend)
+
+    upper_cell_particles = coarse_particles.select(upper_slct_box)
+    lower_cell_particles = coarse_particles.select(lower_slct_box)
+
+    coarse_particles.add(upper_cell_particles.shift_icell(-domain_box.size()))
+    coarse_particles.add(lower_cell_particles.shift_icell(domain_box.size()))
+
+    boxes = {}
+    for ilvl, boxes_data in refinement_boxes.items():
+
+        level_number = int(ilvl.strip("L")) + 1
+
+        if level_number not in boxes:
+            boxes[level_number] = []
+
+        for boxname, lower_upper in boxes_data.items():
+            refinement_box = Box(lower_upper[0][0], lower_upper[1][0])
+            refined_box = boxm.refine(refinement_box, refinement_ratio)
+            boxes[level_number].append(refined_box)
+
+    # coarse level boxes are arbitrarily divided in 2 patches in the middle
+    middle_cell = round(domain_box.upper / 2)
+    lower_box = Box(0, middle_cell)
+    upper_box = Box(middle_cell + 1, domain_box.upper)
+    boxes[0] = [lower_box, upper_box]
+
+    patch_datas = {}
+
+    for ilvl, lvl_box in boxes.items():
+
+        lvl_cell_width = cell_width / (refinement_ratio ** ilvl)
+
+        if ilvl == 0:
+            lvl_particles = coarse_particles
+        else:
+            level_domain_box = boxm.refine(domain_box, refinement_ratio)
+            lvl_ghost_domain_box = boxm.grow(level_domain_box, domain_layout.particleGhostNbr(interp_order))
+            lvl_particles = Particles(box = lvl_ghost_domain_box)
+
+        if ilvl not in patch_datas:
+            patch_datas[ilvl] = []
+
+        for box in lvl_box:
+
+            ghost_box = boxm.grow(box, 5)
```

Comment:
so these are the 5 ghost cells ? I thought it should depend on the interp_order,

Reply:
I answered this already. It does not depend on interp order (hard-coded to 5) for fields. Particles do depend on interporder in the way previously described (1 cell for io 1, 2 cells for 2 and 3)